### PR TITLE
Add first version of websocket extension

### PR DESCRIPTION
**Pull request by JCash**
Supports all platforms, with ws:// and wss://

### CODE REVIEW

File: websocket/src/handshake.cpp [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    {
+        return SetStatus(conn, RESULT_HANDSHAKE_FAILED, "Connection not ready for sending data: %s", dmSocket::ResultToString(sr));
+    }
+
+// In emscripten, the sockets are actually already websockets, so no handshake necessary
+#if defined(__EMSCRIPTEN__)
+    return RESULT_OK;
+#else
+    return SendClientHandshakeHeaders(conn);
+#endif
```
Comment (JCash) [inline websocket/src/handshake.cpp:90]: Some special handling for HTML5

File: websocket/src/handshake.cpp [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+
+    conn->m_BufferSize += recv_bytes;
+
+    // NOTE: We have an extra byte for null-termination so no buffer overrun here.
+    conn->m_Buffer[conn->m_BufferSize] = '\0';
+
+    // Check if the end of the response has arrived
+    if (conn->m_BufferSize >= 4 && strcmp(conn->m_Buffer + conn->m_BufferSize - 4, "\r\n\r\n") == 0)
+    {
+        return RESULT_OK;
```
Comment (JCash) [inline websocket/src/handshake.cpp:151]: Continue reading until we have the final tag (`\r\n\r\n`)

File: websocket/src/handshake.cpp [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+        return SetStatus(conn, RESULT_HANDSHAKE_FAILED, "Missing: '%s' in header", http_version_and_status_protocol);
+    }
+
+    r = strstr(r, "\r\n") + 2;
+
+    bool upgraded = false;
+    bool valid_key = false;
+    const char* protocol = "";
+
+    // TODO: Perhaps also support the Sec-WebSocket-Protocol
```
Comment (JCash) [inline websocket/src/handshake.cpp:180]: We don't yet support all bells and whistles.

File: websocket/src/handshake.cpp [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+    WS_SENDALL("Upgrade: websocket\r\n");
+    WS_SENDALL("Connection: Upgrade\r\n");
+    WS_SENDALL("Sec-WebSocket-Key: ");
+    WS_SENDALL(encoded_key);
+    WS_SENDALL("\r\n");
+    WS_SENDALL("Sec-WebSocket-Version: 13\r\n");
+
+    // Add custom protocols
+
+    // Add custom headers
```
Comment (JCash) [inline websocket/src/handshake.cpp:58]: We could also support extra headers in the future.

File: websocket/src/script_util.cpp [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        if(!lua_isnil(L, -1))
+        {
+            result = (char*)luaL_checkstring(L, -1);
+        }
+        lua_pop(L, 1);
+    }
+    return result;
+}
+
+} // namespace
```
Comment (JCash) [inline websocket/src/script_util.cpp:79]: These are good, and we might want to add them to the engine in the future.

Comment (JCash) [reply websocket/src/script_util.cpp:79]: They are currently unused.
I added them in anticipation of getting arguments from the `options` table to the connect function.

File: websocket/src/socket.cpp [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    int total_sent_bytes = 0;
+    int sent_bytes = 0;
+
+    while (total_sent_bytes < length) {
+        dmSocket::Result r;
+
+        if (conn->m_SSLSocket)
+            r = dmSSLSocket::Send(conn->m_SSLSocket, buffer + total_sent_bytes, length - total_sent_bytes, &sent_bytes);
+        else
+            r = dmSocket::Send(conn->m_Socket, buffer + total_sent_bytes, length - total_sent_bytes, &sent_bytes);
```
Comment (JCash) [inline websocket/src/socket.cpp:27]: I'm quite happy with how clean this code is now :)

File: websocket/src/websocket.cpp [hunk 7]
```diff
@@ -0,0 +1,536 @@
...
+
+static int FindConnection(WebsocketConnection* conn)
+{
+    for (int i = 0; i < g_Websocket.m_Connections.Size(); ++i )
+    {
+        if (g_Websocket.m_Connections[i] == conn)
+            return i;
+    }
+    return -1;
+}
```
Comment (JCash) [inline websocket/src/websocket.cpp:156]: These could be moved to a connection.cpp. But this file isn't too big just yet.

File: websocket/src/websocket.cpp [hunk 8]
```diff
@@ -0,0 +1,536 @@
...
+
+    if (!g_Websocket.m_Initialized)
+        return DM_LUA_ERROR("The web socket module isn't initialized");
+
+    const char* url = luaL_checkstring(L, 1);
+
+    // long playedTime = luaL_checktable_number(L, 2, "playedTime", -1);
+    // long progressValue = luaL_checktable_number(L, 2, "progressValue", -1);
+    // char *description = luaL_checktable_string(L, 2, "description", NULL);
+    // char *coverImage = luaL_checktable_string(L, 2, "coverImage", NULL);
```
Comment (JCash) [inline websocket/src/websocket.cpp:173]: A bit silly to leave here (as documentaiton), but I suspect that we want to have configuration options soon enough.

File: websocket/src/websocket.cpp [hunk 9]
```diff
@@ -0,0 +1,536 @@
...
+        return DM_LUA_ERROR("Invalid connection");
+
+    if (conn->m_State != STATE_CONNECTED)
+        return DM_LUA_ERROR("Connection isn't connected");
+
+    size_t string_length = 0;
+    const char* string = luaL_checklstring(L, 2, &string_length);
+
+#if defined(HAVE_WSLAY)
+    int write_mode = WSLAY_BINARY_FRAME; // WSLAY_TEXT_FRAME
```
Comment (JCash) [inline websocket/src/websocket.cpp:230]: Currently always binary.

File: websocket/src/websocket.cpp [hunk 10]
```diff
@@ -0,0 +1,536 @@
...
+static dmExtension::Result WebsocketFinalize(dmExtension::Params* params)
+{
+    return dmExtension::RESULT_OK;
+}
+
+static dmExtension::Result WebsocketOnUpdate(dmExtension::Params* params)
+{
+    uint32_t size = g_Websocket.m_Connections.Size();
+
+    for (uint32_t i = 0; i < size; ++i)
```
Comment (JCash) [inline websocket/src/websocket.cpp:390]: The state handling.
Upon error, the status flag is set, and the connection is closed.
The cleanup happens next frame, for the `STATE_DISCONNECTED` connections.

File: websocket/src/websocket.cpp [hunk 11]
```diff
@@ -0,0 +1,536 @@
...
+            }
+
+            SetState(conn, STATE_HANDSHAKE_READ);
+        }
+        else if (STATE_CONNECTING == conn->m_State)
+        {
+            dmSocket::Result socket_result;
+            int timeout = g_Websocket.m_Timeout;
+#if defined(__EMSCRIPTEN__)
+            timeout = 0;
```
Comment (JCash) [inline websocket/src/websocket.cpp:514]: As seen in the defold-websocket.

File: websocket/src/websocket.cpp [hunk 12]
```diff
@@ -0,0 +1,536 @@
...
+            {
+                CLOSE_CONN("Failed initializing wslay: %s", WSL_ResultToString(r));
+                continue;
+            }
+
+            dmSocket::SetNoDelay(conn->m_Socket, true);
+            // Don't go lower than 1000 since some platforms might not have that good precision
+            dmSocket::SetReceiveTimeout(conn->m_Socket, 1000);
+            if (conn->m_SSLSocket)
+                dmSSLSocket::SetReceiveTimeout(conn->m_SSLSocket, 1000);
```
Comment (JCash) [inline websocket/src/websocket.cpp:484]: Each frame, the timeout is 1ms.
If this becomes an issue, we'll need to put it on another thread.

File: websocket/src/wslay_callbacks.cpp [hunk 13]
```diff
@@ -0,0 +1,151 @@
+#include "websocket.h"
```
Comment (JCash) [inline websocket/src/wslay_callbacks.cpp:1]: Our wrapper over the wslay library

File: websocket/src/script_util.cpp (hunk 5 above)
Comment (britzl) [reply websocket/src/script_util.cpp:79]: Yes, I think there's a bunch of helper functions for checking Lua types and returning default values that are used in several extensions. Perhaps also things such as pushing a key-value pair on a table.

File: websocket/src/socket.cpp (hunk 6 above)
Comment (britzl) [reply websocket/src/socket.cpp:27]: I agree. It looks great!

**Review by britzl — APPROVED** [review]
Pretty big PR but from what I can tell it looks good!

**Pull request merged**